boa/src/builtins/function/mod.rs

```diff
@@ -260,7 +257,7 @@ impl Function {
 
                 // Call body should be set before reaching here
                 let _ = match &self.body {
-                    FunctionBody::Ordinary(ref body) => interpreter.run(body),
+                    FunctionBody::Ordinary(ref body) => body.run(interpreter),
```

Comment:
I new environment was already created with `this` set.
body.run calls StatementList which sets a `declarativeEnvironment` again, so `this` is lost.